### ISSUE TITLE
CRF: Incompatible shapes

### ISSUE THREAD

**Issue by grayfall** [farizrahman4u/keras-contrib#160]
[stack_trace.txt](https://github.com/farizrahman4u/keras-contrib/files/1366463/stack_trace.txt)
I've been playing with the CRF layer contributed by @linxihui. I've written a fairly simple RNN-CRF to get a taste of the API:

```
from keras import layers, models
from keras_contrib.layers import CRF

maxlen = 200
embed = 50
x = ...   # a zero-padded array of word-embeddings; shape = (nsamples, maxlen, embed)
y = ...   # a zero-padded array of target labels encoded as indices; shape = (nsamples, maxlen)
masks = ...  # boolean padding masks; shape = (nsamples, maxlen)

l_in = layers.Input((maxlen, embed))

rnn = layers.GRU(200, return_sequences=True)(l_in)
crf = CRF(2, sparse_target=True)
l_out = crf(rnn)

model = models.Model(l_in, l_out)
model.compile(optimizer="Adam", loss=crf.loss_function, 
              metrics=[crf.accuracy],
              sample_weight_mode="temporal")

model.fit(x, y[:,:,None], sample_weight=masks.astype(np.float32),
                verbose=1, epochs=5, batch_size=100)
```
Shortly after the training initialises, I get the following error

```
InvalidArgumentError (see above for traceback): Incompatible shapes: [100] vs. [100,200]
	 [[Node: loss_3/crf_5_loss/mul_2 = Mul[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"](loss_3/crf_5_loss/Mean, _arg_crf_5_sample_weights_0_1/_327)]]
	 [[Node: metrics_3/acc/while_1/transpose/_365 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_1393_metrics_3/acc/while_1/transpose", tensor_type=DT_INT32, _device="/job:localhost/replica:0/task:0/cpu:0"](^_cloopmetrics_3/acc/while_1/ExpandDims/dim/_313)]]
```

Seems like it can't correctly infer that `y` is a 2D array. I also tried converting `y` into one-hot encoded annotations while passing `sparse_target=False` to the CRF layer, but that gives me the same error. There is no difference between the sequential and functional APIs, too. At the same time, everything works fine if I replace the CRF output layer with

```
l_out = layers.TimeDistributed(layers.Dense(1, activation="sigmoid"))(rnn)
```
Am I doing something wrong or is this a bug? My setup:
```
Keras==2.0.8
keras-contrib==1.2.1
tensorflow==1.3.0
```
I'm attaching the complete stack trace, too.

**Comment by TimRepke** [comment]
Hi @grayfall ,
I just stumbled across the exact same issue! What I found is, that the CRF layer, or it's loss function rather[1], doesn't support sample weights. I removed the weights and `sample_weight_mode="temporal"`, which made it work.

In your example, you use weights for masking as far as I can tell. The CRF Layer documentation says:
> ### Masking
>    This layer supports masking for input data with a variable number
>    of timesteps. To introduce masks to your data,
>    use an [Embedding](embeddings.md) layer with the `mask_zero` parameter
>    set to `True`.

Hope that helps
Tim

[1] https://github.com/farizrahman4u/keras-contrib/blob/master/keras_contrib/layers/crf.py#L284

**Comment by linxihui** [comment]
@grayfall  The default training mode is to minimize the negative log likelihood of join distribution which is how the CRF is defined.  You cannot assign weights on each time slice, and thus `sample_weight_mode="temporal"` is not allowed.   If you want to really needs that, go with *marginal mode* for training, which minimize the usual categorical entropy. It is not exactly the objective for CRF but an approximation.

It seems all you need is to *mask* you time steps, not really a weighting. If so, you don't need pass it as sample weight, but rather as Keras masking, either by `Embedding(..., mask_zero=True)` or `Masking` layer, or write your own layer to manually set your mask to the input tensor before CRF.

**Comment by grayfall** [comment]
@TimRepke @linxihui Thank you for the reply. I actually need to pass the weights to my models, because they contain convolutional layers, which as of this writing do not support Keras' built-in masking, but I guess I'll figure something out.
